File: MenuItemTemplateCS/DataSource/BaseDataSource.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Runtime;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;
using Corel.Interop.VGCore;

namespace $safeprojectname$.DataSource
{
    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.AutoDual)]
    public class BaseDataSource : INotifyPropertyChanged
    {
        protected DataSourceProxy m_AppProxy;

        public BaseDataSource(DataSourceProxy proxy)
        {
            this.m_AppProxy = proxy;
        }

        public event PropertyChangedEventHandler PropertyChanged;
        
        public void NotifyPropertyChanged([CallerMemberName]string propertyName = "")
        {
            try
            {
                this.m_AppProxy.UpdateListeners(propertyName);
                if (PropertyChanged != null)
                    PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
            catch { }
        }
    }
}

File: MenuItemTemplateVB/DataSource/BaseDataSource.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Runtime
Imports System.Runtime.CompilerServices
Imports System.Runtime.InteropServices
Imports System.Text
Imports Corel.Interop.VGCore

Namespace DataSource

    <ComVisible(True)>
    <ClassInterface(ClassInterfaceType.AutoDual)>
    Public Class BaseDataSource
        Implements INotifyPropertyChanged

        Protected m_AppProxy As DataSourceProxy

        Public Sub New(ByVal proxy As DataSourceProxy)
            Me.m_AppProxy = proxy
        End Sub

        Public Event PropertyChanged As PropertyChangedEventHandler
        Private Event INotifyPropertyChanged_PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Public Sub NotifyPropertyChanged(
            <CallerMemberName> ByVal Optional propertyName As String = "")
            Try
                Me.m_AppProxy.UpdateListeners(propertyName)
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
            Catch
            End Try
        End Sub

    End Class
